File: Gold이상문제_정리/파티_1238.java
package Gold이상문제_정리;

import java.io.*;
import java.util.*;

// 학생의 시작점 - > 도착점 -> 시작점의 총 소요시간을 계산하고 그중 가장 오래 걸리는 학생의 시간을 출력

public class 파티_1238 {
	static PriorityQueue<Road> pq = new PriorityQueue<>(Comparator.comparingInt(s -> s.time));
	static int[] dp;
	static ArrayList<Road>[] adj;
	static int N, M, X;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		X = Integer.parseInt(st.nextToken());

		adj = new ArrayList[N+1];
		for (int i = 1; i <= N; i++)
			adj[i] = new ArrayList<>();

		for (int i = 0; i < M; i++)
		{
			st = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(st.nextToken());
			int e = Integer.parseInt(st.nextToken());
			int t = Integer.parseInt(st.nextToken());
			adj[s].add(new Road(e, t));
		}

		int MAX_TIME = 0;
		for (int i = 1; i <= N; i++) {
			// 가는데 비용
			pq.clear();
			pq.add(new Road(i, 0));
			start(i);
			int go = dp[X];

			// 오는데 비용
			pq.clear();
			pq.add(new Road(X, 0));
			start(X);
			int back = dp[i];
			MAX_TIME = Math.max(MAX_TIME, go + back);
		}
		System.out.println(MAX_TIME);
	}

	private static void start(int start) {
		boolean[] visited = new boolean[N+1];
		dp = new int[N+1];
		Arrays.fill(dp, Integer.MAX_VALUE);
		dp[start] = 0;

		while (!pq.isEmpty())
		{
			Road now = pq.poll();

			visited[now.end] = true;

			if (dp[now.end] < now.time) continue;

			for (Road next : adj[now.end])
			{
				if (!visited[next.end] && dp[now.end] + next.time < dp[next.end])
				{
					dp[next.end] = dp[now.end] + next.time;
					pq.add(new Road(next.end, dp[next.end]));
				}
			}
		}
	}

	private static class Road
	{
		int end;
		int time;

		public Road(int end, int time) {
			this.end = end;
			this.time = time;
		}
	}
}

File: 다익스트라/파티_1238.kt
package 다익스트라

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.Arrays
import java.util.PriorityQueue
import java.util.StringTokenizer
import kotlin.math.max

/*
4 8 2
1 2 4
1 3 2
1 4 7
2 1 1
2 3 5
3 1 2
3 4 4
4 2 3
 */

private var N = 0
private var M = 0
private var X = 0
private const val MAX = 987654321
private var adj = Array(N + 1) { mutableListOf<Vertex>() }
private var dp = IntArray(N + 1) { MAX }
private var visited = Array(N + 1) { false }

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    N = st.nextToken().toInt()
    M = st.nextToken().toInt()
    X = st.nextToken().toInt()

    adj = Array(N + 1) { mutableListOf() }
    dp = IntArray(N + 1) { MAX }
    visited = Array(N + 1) { false }

    repeat(M) {
        st = StringTokenizer(br.readLine())
        val from = st.nextToken().toInt()
        val to = st.nextToken().toInt()
        val cost = st.nextToken().toInt()
        adj[from].add(Vertex(to, cost))
    }

    var maxTime = 0
    for (now in 1..N) {
        dijkstra(now)
        val goTime = dp[X]
        dijkstra(X)
        val backTime = dp[now]
        maxTime = max(maxTime, goTime + backTime)
    }
    print(maxTime)
}

private fun dijkstra(start: Int) {
    dp.fill(MAX)
    visited.fill(false)
    val pq = PriorityQueue<Vertex>(compareBy { it.cost })
    pq.add(Vertex(start, 0))
    dp[start] = 0

    while (pq.isNotEmpty()) {
        val (now, cost) = pq.poll()

        if (dp[now] < cost) continue

        visited[now] = true

        for (next in adj[now]) {
            if (!visited[next.to] && dp[now] + next.cost < dp[next.to]) {
                dp[next.to] = dp[now] + next.cost
                pq.add(Vertex(next.to, dp[next.to]))
            }
        }
    }
}

private data class Vertex(
    val to: Int,
    val cost: Int,
)